src/box/lua/config/applier/box_cfg.lua

```diff
@@ -754,6 +787,10 @@ local function apply(config)
             box_cfg.read_only = true
         end
 
+        -- RO may be enforced by the isolated mode, so we call the
+        -- function after all the other RO/RW logic is done.
+        switch_isolated_mode(configdata, box_cfg)
+
```

Comment:
I don't like the same `switch_isolated_mode()` in two parts though it seems safer than any other approach. Also it brings kinda:
```
force_ro_if_some_condition()
forcier_ro_if_some_condition()
```

If resulting `needs_retry` can be `false` if the instance is isolated seems like there is no difference between startup and non startup scenarios and switch to ro can be done before all of these checks.
```
if isolated then <enforce ro> return end

if is_startup and <other conds> then 
  <old startup code>
end

<otherwise code>
```

Have you considered this idea?

Reply:
Thank you for refactoring and for a detailed description of the suggested approach.